cpp/src/arrow/builder.cc

```diff
@@ -1403,4 +1404,87 @@ Status MakeDictionaryBuilder(MemoryPool* pool, const std::shared_ptr<DataType>&
   }
 }
 
+#define DICTIONARY_ARRAY_CASE(ENUM, BuilderType)                           \
+  case Type::ENUM:                                                         \
+    builder = std::make_shared<BuilderType>(pool, type);                   \
```

Comment:
This is using a noew deprecated form of this constructor, can you flip the argument order?